### PR TITLE
Improvements in the reader from openPMD file

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR proposes the following improvements, all meant to achieve #127:
- The option to output the normalized vector potential instead of the electric field is implemented. For the rationale, see https://github.com/openPMD/openPMD-standard/pull/281. This is certainly open for discussion. Converters from E to a0 and reverse are implemented in utils (it is a very vanilla version, definitely room for improvement there).
- The calculation of local and average frequency is in a separate function. This way, we can implement different methods to do this operation (e.g. FFT-based) easily.
- profile from_array and from_openPMD now also support RZ representation. Without this, operating on the 3D data was not practical. Again, very vanilla implementation, the implementation assumes axial symmetry for the envelope, the mode decomposition is not included in this PR.

With these changes, the output of an FBPIC simulation can be sent as the input of a HiPACE++ simulation. The figure below shows, for a laser pulse in a parabolic channel, the pulse width during propagation, when starting the HiPACE++ simulation from the output of the FBPIC simulation at different times. Significant differences can be observed, which could very much come from differences between the two codes: even without using LASY the two codes give slightly different results for this case. This is probably a matter of tuning the resolution, box size, time step etc., so I do not think it is necessarily a problem.
laser_width_hipace_fbpic
[laser_width_hipace_fbpic.pdf](https://github.com/LASY-org/lasy/files/12252598/laser_width_hipace_fbpic.pdf)

The scripts I used are posted below, so we can later work on the agreement. I had to add some .txt to upload the files.

[inputs_fbpic.py.txt](https://github.com/LASY-org/lasy/files/12252686/inputs_fbpic.py.txt)
[inputs_hipace.txt](https://github.com/LASY-org/lasy/files/12252687/inputs_hipace.txt)
[script_fbpic.sh.txt](https://github.com/LASY-org/lasy/files/12252689/script_fbpic.sh.txt)
[script_hipace.sh.txt](https://github.com/LASY-org/lasy/files/12252690/script_hipace.sh.txt)

The lasy handling was done using:
```python
from lasy.profiles.from_openpmd_profile import FromOpenPMDProfile
from lasy.laser import Laser
import matplotlib.pyplot as plt
from openpmd_viewer.addons import LpaDiagnostics
import scipy.constants as scc

profile2 = FromOpenPMDProfile('./DATA', 26, (0,1), 'E', 'y', theta=0)

dim = 'xyt'
lo = (-100e-6, -100e-6, 0)
hi = (+100e-6, +100e-6, (m0.zmax-m0.zmin)/scc.c)
npoints=(128, 128, 500)
## Tested, also works with
# dim = 'rt'
# lo = (0, 0)
# hi = (+100e-6, (m0.zmax-m0.zmin)/scc.c)
# npoints=(300, 600)
laser = Laser(dim, lo, hi, npoints, profile2)
laser.write_to_file('./lasy_t3/laser3d', use_a0=True)
```

### CODE REVIEW

**Comment by MaxThevenet** [comment]
Tests pass, this is ready for review :)

File: lasy/profiles/from_openpmd_profile.py [hunk 1]
```diff
@@ -43,56 +45,171 @@ class FromOpenPMDProfile(FromArrayProfile):
...
+                omg0 = it.meshes["laserEnvelope"].get_attribute("angularFrequency")
+                wavelength = 2 * np.pi * c / omg0
+                array = F
+
+            axes_order = ["r", "t"]
+
+        print(
+            "Wavelength used in the definition of the envelope (nm):",
+            wavelength * 1.0e9,
+        )
```
Comment (AngelFP) [inline lasy/profiles/from_openpmd_profile.py]: Could we avoid this print by default? maybe with a `verbose` parameter.

File: lasy/utils/laser_utils.py [hunk 2]
```diff
@@ -205,3 +207,157 @@ def get_full_field(laser, theta=0, slice=0, slice_axis="x", Nt=None):
...
+    else:
+        assert dim in ["xyt", "rt"]
+        if dim == "xyt" and not phase_unwrap_1d:
+            print("WARNING: using 3D phase unwrapping, this can be expensive")
+
+        if not is_hilbert:
+            h = np.squeeze(hilbert(grid.field))
+        else:
+            h = np.squeeze(grid.field)
+        print(h.shape)
```
Comment (AngelFP) [inline lasy/utils/laser_utils.py]: Probably a leftover print from debugging

File: lasy/profiles/from_openpmd_profile.py (hunk 1 above)
Comment (MaxThevenet) [reply lasy/profiles/from_openpmd_profile.py]: Thanks! This should be fixed

**Comment by AngelFP** [comment]
Results from a similar test restarting an FBPIC simulation with Wake-T. Looking promising so far!

image

**Comment by MaxThevenet** [comment]
Thanks @AngelFP. The laser propagation is quite well-captured. Final iteration below. Unfortunately, the agreement still isn't great, but nothing that can obviously be attributed to LASY, probably some convergence question.
fig

File: lasy/profiles/from_openpmd_profile.py [hunk 3]
```diff
@@ -43,56 +45,176 @@ class FromOpenPMDProfile(FromArrayProfile):
...
-        else:
-            t = m.t
-        axes = {"x": m.x, "y": m.y, "t": t}
-
-        # If array does not contain the envelope but the electric field,
-        # extract the envelope with a Hilbert transform
-        if envelope == False:
-            # Assumes z is last dimension!
-            h = hilbert(F)
+        F, m = ts.get_field(iteration=iteration, field=field, coord=coord, theta=theta)
```
Comment (MaxThevenet) [inline lasy/profiles/from_openpmd_profile.py:81]: In principle, we could avoid depending on the viewer here, but that would take a bit more work (in particular to properly convert from RZ decomposition of fields to RZ decomposition of envelope). I would be happy to discuss this.

File: lasy/laser.py [hunk 4]
```diff
@@ -398,7 +398,7 @@ def import_from_z(self, field_z, z_axis, z0=0.0, t0=0.0, backend="NP"):
             self.grid.field = prop.z2t(transform_data, t_axis, z0=z0, t0=t0).T
             self.grid.field *= np.exp(1j * (z0 / c + t_axis) * omega0)
 
-    def write_to_file(self, file_prefix="laser", file_format="h5"):
+    def write_to_file(self, file_prefix="laser", file_format="h5", use_a0=False):
         """
         Write the laser profile + metadata to file.
```
Comment (AngelFP) [inline lasy/laser.py:406]: Could you add this to the docstring here too
```
use_a0 : bool (optional)
    Whether the envelope is converted to normalized vector potential
    before writing to file.
```

Also, perhaps `a0` can be a bit confusing (i.e., as the maximum of a), maybe something like `use_a` or more explicit like `save_as_potential`.

File: lasy/profiles/from_openpmd_profile.py [hunk 5]
```diff
@@ -43,56 +45,176 @@ class FromOpenPMDProfile(FromArrayProfile):
         Prefix of the openPMD file from which the envelope is read.
         Only used when envelope=True.
         The provided iteration is read from <path>/<prefix>_%T.h5.
+
```
Comment (AngelFP) [inline lasy/profiles/from_openpmd_profile.py:48]: There's one vertical space too much here

File: lasy/profiles/from_openpmd_profile.py [hunk 6]
```diff
@@ -43,56 +45,176 @@ class FromOpenPMDProfile(FromArrayProfile):
...
+            else:
+                t = m.t
+            axes = {"x": m.x, "y": m.y, "t": t}
+
+            if phase_unwrap_1d is None:
+                phase_unwrap_1d = True
+
+            # If array does not contain the envelope but the electric field,
+            # extract the envelope with a Hilbert transform
+            if envelope == False:
```
Comment (AngelFP) [inline lasy/profiles/from_openpmd_profile.py]:
```suggestion
            if not envelope:
```

File: lasy/utils/laser_utils.py [hunk 7]
```diff
@@ -205,3 +207,165 @@ def get_full_field(laser, theta=0, slice=0, slice_axis="x", Nt=None):
...
+
+    # Filter out too small frequencies
+    omega = np.maximum(omega, lower_bound * central_omega)
+    # Filter out too large frequencies
+    omega = np.minimum(omega, upper_bound * central_omega)
+
+    return omega, central_omega
+
+
+def field_to_a0(grid, omega0):
```
Comment (AngelFP) [inline lasy/utils/laser_utils.py]: Same comment as above. Maybe `a0` can be misleading. `field_to_potential` could also work, for example. Not a big issue in any case.

**Review by AngelFP — APPROVED** [review]

File: lasy/profiles/from_openpmd_profile.py (hunk 5 above)
Comment (MaxThevenet) [reply lasy/profiles/from_openpmd_profile.py:48]: Thanks! I fixed it.

File: lasy/profiles/from_openpmd_profile.py (hunk 6 above)
Comment (MaxThevenet) [reply lasy/profiles/from_openpmd_profile.py]: fixed

**Comment by MaxThevenet** [comment]
Thanks @AngelFP for your review, I addressed your comments in the last commits.

**Comment by AngelFP** [comment]
Great, thanks!

**Pull request merged**